test/e2e/utils/imageregistry.go

```diff
@@ -0,0 +1,181 @@
+		e2e.DeleteDockerRegistry(c, testNamespace)
+	}
+	return nil
+}
+
+
+// podCheckFunc describes a function that returns true if the given Pod meets some criteria; false otherwise.
+type podCheckFunc func(pod *corev1.Pod) bool
+
+func awaitPod(c operatorclient.ClientInterface, namespace, name string, checkPod podCheckFunc) (*corev1.Pod, error) {
+	var pod *corev1.Pod
+	err := wait.Poll(pollInterval, pollDuration, func() (bool, error) {
+		p, err := c.KubernetesInterface().CoreV1().Pods(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		pod = p
+		return checkPod(pod), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+func Local(client operatorclient.ClientInterface) (bool, error) {
+	const ClusterVersionGroup = "config.openshift.io"
+	const ClusterVersionVersion = "v1"
+	const ClusterVersionKind = "ClusterVersion"
+	gv := metav1.GroupVersion{Group: ClusterVersionGroup, Version: ClusterVersionVersion}.String()
+
+	groups, err := client.KubernetesInterface().Discovery().ServerResourcesForGroupVersion(gv)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return true, fmt.Errorf("checking if cluster is local: checking server groups: %s", err)
+	}
+
+	for _, group := range groups.APIResources {
+		if group.Kind == ClusterVersionKind {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// podReady returns true if the given Pod has a ready condition with ConditionStatus "True"; false otherwise.
+func podReady(pod *corev1.Pod) bool {
+	var status corev1.ConditionStatus
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type != corev1.PodReady {
+			// Ignore all condition other than PodReady
+			continue
+		}
+
+		// Found PodReady condition
+		status = condition.Status
+		break
+	}
+
+	return status == corev1.ConditionTrue
+}
+
+
+// Recreates index each time, does not update index. Returns the indexReference to use in CatalogSources
+func (r *Registry) CreateBundlesAndIndex(indexName string, bundles []*Bundle) (string, error) {
+	bundleRefs := make([]string, 0)
+
+	switch r.BundleTool {
+	case DockerTool:
+		for _, b := range bundles {
+			ref, err := r.buildBundleImage(b)
+			if err != nil {
+				return "", err
+			}
+			bundleRefs = append(bundleRefs, ref)
+		}
+
+		if err := r.uploadBundleReferences(bundleRefs); err != nil {
+			return "", err
+		}
+	case BuildahTool:
+		for _, b := range bundles {
+			labels := generateBundleLabels(b.PackageName, b.DefaultChannel, b.Channels)
+			imageRef, err := createLocalBundleImage(path.Dir(b.BundleManifestDirectory), b.BundlePath, []string{"manifests", "metadata"}, labels)
+			if err != nil {
+				return "", err
+			}
+			if err := r.skopeoCopy(fmt.Sprintf("%s/%s:", r.URL, b.BundlePath), b.Version, imageRef, ""); err != nil {
+				return "", err
+			}
+			bundleRefs = append(bundleRefs, imageRef)
+		}
+	}
+
+	indexReference, err := r.CreateAndUploadIndex(indexName, bundleRefs)
```

Comment:
one gotcha here is that we need to add the built bundle images to the index in the correct order - if one bundle is after a prior bundle, but the CSV has a `replaces` reference to a CSV that is not yet in the index - opm will error out.

I think we did some work to address this (the ability to add multiple bundles, not necessarily in the right order), but not sure if it's still an issue. Locally I ran into this issue recently. @kevinrizza this is still an issue right? 

tl;dr we should be sure we are adding the bundles in the correct order

Reply:
We're using the multiple bundles in one go approach here, each time it creates a new index. I had issues with bundle order when adding one at a time, and this seemed quicker than a build-upload loop, so switched to it.